### ISSUE TITLE
Umlauts in Search lead to JS-Error.

### ISSUE THREAD

**Issue by animaux** [mazedigital/association_ui_selector#55]
When using the search term `schö` for example an error is thrown and the search is not successful:

    TypeError: undefined is not an object (evaluating 'a.length')

**Comment by animaux** [comment]
This results in Strange behaviour:

If I try to look for an entry named `Hänigsen` the above error will prevent a successful search as soon as the »ä« comes into play.

However, if I type `nigsen`, the entry is succefully found and presented. After that I can even start a new search with the search term `Hän…` actually working.

I suspect there is some encoding problem happening during typing or search submission. I wonder if @nilshoerrmann has has this problem too? :)

**Comment by animaux** [comment]
Not sure if the extension or symphony is at fault.

![bildschirmfoto 2016-12-14 um 09 13 19](https://cloud.githubusercontent.com/assets/446874/21174465/9615e60e-c1dd-11e6-9e5e-5ace520ddfb3.png)

**Comment by nilshoerrmann** [comment]
No, I haven't encountered this issues yet.

**Comment by animaux** [comment]
Hmpfdgnf. Danke :)

**Comment by animaux** [comment]
Apparently the problem seems to be with Umlauts **in the first 3 letters** of a search term.

**Comment by animaux** [comment]
Version `1.1.1` seems to work fine.

**Comment by animaux** [comment]
Using a reflection field instead of a text field for the association makes me able to find the Umlaut-Entries but still throws the error.

**Comment by animaux** [comment]
Hmmm. Can’t reproduce the error on a copy of the SOR site trying several versions of AI Editor and the Association field itself. Must be some other extension gretching in.

**Comment by nilshoerrmann** [comment]
The search is powered by Selectize that is bundled with Symphony (or at least was – I'm not sure how this is handled in the latest releases). Maybe Selectize was updated in the core resulting in compatibility issues?

**Comment by animaux** [comment]
I have used an updated version of selectize in my fork-set that overrides the bundled one:

https://github.com/animaux/association_ui_selector
https://github.com/animaux/association_ui_editor

Installing those does not break the SOR-Site either … must be something else in play.

**Comment by animaux** [comment]
Really beats me. I’ve copied all the extensions, symphony-folder and rebuilt the section/link-structure from a site with the error but still can’t reproduce it. ¯\_(ツ)_/¯

**Comment by nilshoerrmann** [comment]
Have you tried another browser? 
Have you checked which line is causing the error?

**Comment by animaux** [comment]
![bildarchiv archivrh de](https://cloud.githubusercontent.com/assets/446874/21179361/056173a0-c1f4-11e6-9e2a-c7782299a2b3.png)
![goldwiege de](https://cloud.githubusercontent.com/assets/446874/21179359/04c0250e-c1f4-11e6-89ab-ff40a4e5eb93.png)
![fotoarchiv animaux de](https://cloud.githubusercontent.com/assets/446874/21179358/047123d2-c1f4-11e6-8de2-24eb937e0635.png)

**Comment by animaux** [comment]
And Firefox Dev:

![bildschirmfoto 2016-12-14 um 11 55 16](https://cloud.githubusercontent.com/assets/446874/21179407/3742f6e6-c1f4-11e6-9409-6832cf5f7b67.png)

**Comment by animaux** [comment]
Not really helping because the files mentioned are identical on two sites. One has the error. One not.

**Comment by animaux** [comment]
Extensions are identical too. Site structure. I’m struggling to find the difference …

**Comment by twiro** [comment]
I don't think the problem is caused by the extension itself - I remember having some filtering trouble with umlauts in the publish filtering feature (which also relies on selectize) too. And as far as I can remember I was getting similar javascript-errors as the ones you're posting.

I did some [tests](https://github.com/twiro/symphony-filtering-tests) back then, compared datasource filtering with publish filtering and noticed that there are really a **lot** of quirks regarding filtering umlauts and special characters. Regarding selectize I even noticed differences in typing a character and using the proposed autocomplete-values.

You can see the [results for the basic input-field here](https://github.com/twiro/symphony-filtering-tests/blob/master/fields/input-field.md). 

I submitted a PR that fixes the filtering-bugs number-field has (will be in 2.7), but didn't dig deeper into the umlaut-issues...

**Comment by animaux** [comment]
Thanks Roman.

**Comment by nilshoerrmann** [comment]
That's a great test, Roman. 
You should notify @nitriques and @brendo about the results (if you haven't done this already).

**Comment by twiro** [comment]
> _That's a great test, Roman._

Thank you! I got inspired by [this page](https://github.com/symphonycms/symphony-2/wiki/DateField) in the wiki and thought it would be cool to have this kind of overview for all (core) fields.

> _You should notify @nitriques and @brendo about the results (if you haven't done this already)._

[You bet I did](https://github.com/symphonycms/symphony-2/issues/2564) - but I guess there are no resources left at the moment to take care of things like these. Though  these documented tests would probably be of great help if someone really wanted to start fixing these bugs one by one... as proposed in the post linked above one could start to file separate issues for the separate bugs, but I didn't want to flood symphony's issue tracker with a ton of "edge case issues" that probably won't get fixed for a looong time.

But I do hope to get back to these tests some day, update them and maybe try fixing some of the more problematic issues myself. I also think it would be a great addition to some extensions to directly add these filtering-test-reports to the extension repository ([the way I did with slider field](https://github.com/twiro/slider/blob/master/tests.md)) - helps giving a better understanding of what each field's filtering capabilities (and shortcomings/bugs) are.

But for now... enough issue-hijacking ;)

**Comment by animaux** [comment]
> But for now... enough issue-hijacking ;)

No worries, good thing you did.

I’ve checked the DB encoding too. Same.

**Comment by animaux** [comment]
`.htaccess` and `config.php` are also not the problem …

**Comment by nitriques** [comment]
@animaux curious to see how Chrome handles it. Same error ? I know for a fact that Safari uses UTF-16 in their jscore implementation. But seeing that the error is in jquery itself puzzle me a lot. But again, PHP has not internal encoding (it's just binary data) but you can "format" that binary code to whatever you want.

Are the sites both on the same server ? Same php.ini file ?

**Comment by nitriques** [comment]
@twiro 

> but I guess there are no resources left at the moment to take care of things like these. 

Yeah, sadly. We got "big" clients this year which means we are doing more maintenance that what we've planned, which reduced my "free" time, both at work and at home. But dealing with encoding always was and still is a pain. I wonder how other systems handle those cases, but from my experience, it's not really better.

**Comment by nitriques** [comment]
And this is worth reading: http://softwareengineering.stackexchange.com/a/151296

> strlen is encoding-agnostic and assumes a fixed-width 8 bit encoding, that is, one byte per character, so it will count bytes, not characters.

Make sure to use `General::strlen` and have mb_string installed.

**Comment by animaux** [comment]
> Are the sites both on the same server ? Same php.ini file ?
Yes, absolutely.

And it’s the same in chrome:

![skjermbilde 2016-12-16 kl 19 18 52](https://cloud.githubusercontent.com/assets/446874/21273579/8abf62ba-c3c4-11e6-912f-ccb2a8a9caac.png)

**Comment by animaux** [comment]
I’m not sure entirely it’s an encoding problem since the error comes up as well if I type in at least three standard ascii characters. But in that case the search seems to work.

**Comment by animaux** [comment]
@jonmifsud’s suggestion helped solve the Umlaut-issue:

> change line 12 of association_ui_selector/content/content.query.php to
> `$search = General::sanitize(urldecode($_GET['query']));`

About the error:

> in line 215 of the JS there should be a check if result.entries is populated

So that's the basic recipe to fix this problem :)

**Comment by animaux** [comment]
Fixed in https://github.com/mazedigital/association_ui_selector/pull/57
